File: src/Net/1. First Course/1. Initial Tee/RefactoringGolf.Store/Order.cs
namespace RefactoringGolf.Store
{
    using System;
    using System.Collections.Generic;

    public class Order
    {
        public Customer Customer { get; private set; }

        public Salesman Salesman { get; private set; }

        public DateTime OrderedOn { get; private set; }

        public string DeliveryStreet { get; private set; }

        public string DeliveryCity { get; private set; }

        public string DeliveryCountry { get; private set; }

        public IList<OrderItem> Items { get; private set; }

        public Order(Customer customer, Salesman salesman, string deliveryStreet, string deliveryCity, string deliveryCountry, DateTime orderedOn)
        {
            Customer = customer;
            Salesman = salesman;
            DeliveryStreet = deliveryStreet;
            DeliveryCity = deliveryCity;
            DeliveryCountry = deliveryCountry;
            OrderedOn = orderedOn;
            Items = new List<OrderItem>();
        }

        public decimal Total()
        {
            decimal totalItems = 0;
            foreach (var item in this.Items)
            {
                decimal itemAmount = item.Product.UnitPrice * item.Quantity;
                if (item.Product.Category == ProductCategory.Accessories)
                {
                    decimal booksDiscount = 0;
                    if (itemAmount >= 100)
                    {
                        booksDiscount = itemAmount * 10 / 100;
                    }
                    itemAmount = itemAmount - booksDiscount;
                }
                if (item.Product.Category == ProductCategory.Bikes)
                {
                    // 20% discount for Bikes
                    itemAmount = itemAmount - itemAmount * 20 / 100;
                }
                if (item.Product.Category == ProductCategory.Cloathing)
                {
                    decimal cloathingDiscount = 0;
                    if (item.Quantity > 2)
                    {
                        cloathingDiscount = item.Product.UnitPrice;
                    }
                    itemAmount = itemAmount - cloathingDiscount;
                }
                totalItems += itemAmount;
            }

            if (this.DeliveryCountry == "USA")
            {
                //total=totalItems + tax + 0 shipping
                return totalItems + totalItems * 5 / 100;
            }

            //total=totalItems + tax + 15 shipping
            return totalItems + totalItems * 5 / 100 + 15;
        }
    }
}

File: src/Java/1. First Course/1. Initial Tee/src/main/refactoringgolf/store/Order.java
package refactoringgolf.store;

import java.util.Date;
import java.util.HashSet;
import java.util.Set;

public class Order {

	private Customer customer;
	private Salesman salesman;
	private Date orderedOn;
	private String deliveryStreet;
	private String deliveryCity;
	private String deliveryCountry;
	private Set<OrderItem> items;

	public Order(Customer customer, Salesman salesman, String deliveryStreet, String deliveryCity, String deliveryCountry, Date orderedOn) {
		this.customer = customer;
		this.salesman = salesman;
		this.deliveryStreet = deliveryStreet;
		this.deliveryCity = deliveryCity;
		this.deliveryCountry = deliveryCountry;
		this.orderedOn = orderedOn;
		this.items = new HashSet<OrderItem>();
	}

	public Customer getCustomer() {
		return customer;
	}

	public Salesman getSalesman() {
		return salesman;
	}

	public Date getOrderedOn() {
		return orderedOn;
	}

	public String getDeliveryStreet() {
		return deliveryStreet;
	}

	public String getDeliveryCity() {
		return deliveryCity;
	}

	public String getDeliveryCountry() {
		return deliveryCountry;
	}

	public Set<OrderItem> getItems() {
		return items;
	}

	public float total() {
		float totalItems = 0;
		for (OrderItem item : items) {
			float totalItem=0;
			float itemAmount = item.getProduct().getUnitPrice() * item.getQuantity();
			if (item.getProduct().getCategory() == ProductCategory.Accessories) {
				float booksDiscount = 0;
				if (itemAmount >= 100) {
					booksDiscount = itemAmount * 10 / 100;
				}
				totalItem = itemAmount - booksDiscount;
			}
			if (item.getProduct().getCategory() == ProductCategory.Bikes) {
				// 20% discount for Bikes
				totalItem = itemAmount - itemAmount * 20 / 100;
			}
			if (item.getProduct().getCategory() == ProductCategory.Cloathing) {
				float cloathingDiscount = 0;
				if (item.getQuantity() > 2) {
					cloathingDiscount = item.getProduct().getUnitPrice();
				}
				totalItem = itemAmount - cloathingDiscount;
			}
			totalItems += totalItem;
		}

		if (this.deliveryCountry == "USA"){
			// total=totalItems + tax + 0 shipping
			return totalItems + totalItems * 5 / 100;
		}

		// total=totalItemst + tax + 15 shipping
		return totalItems + totalItems * 5 / 100 + 15;
	}
}
